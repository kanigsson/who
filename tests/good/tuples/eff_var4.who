type unit
logic tt : unit
type region ['u]
type refty ['reg 'ty]
logic ref_get ['reg 'u] : 'reg region -> ('reg,'u) refty -> 'u

logic get_2_1_tuple ['a 'b] : 'a * 'b -> 'a
logic get_2_2_tuple ['a 'b] : 'a * 'b -> 'b
logic mk_2tuple ['a 'b] : 'a -> 'b -> 'a * 'b


logic = ['a] : 'a -> 'a -> prop
logic -> : prop -> prop -> prop
logic !! ['a|r|'e] : ref(r,'a) -> <r 'e > -> 'a

logic z : int
letregion b
logic f [||'e] :  <b 'e> -> int
goal g :
  forall (x : ref(b,int)) (u : <b>).
    z = f[{}] u -> !!x|u = 3
