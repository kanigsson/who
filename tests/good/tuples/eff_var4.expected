type unit
logic tt  : unit
type region['u||]
type refty['reg 'ty||]
logic ref_get ['reg 'u||] : region[reg] -> refty[reg,u] -> u
logic get_2_1_tuple ['a 'b||] : (a * b) -> a
logic get_2_2_tuple ['a 'b||] : (a * b) -> b
logic mk_2tuple ['a 'b||] : a -> b -> a * b
logic = ['a||] : a -> a -> prop
logic ->  : prop -> prop -> prop
logic z  : int
INTROS ['b||]
logic f ['e||] : (region[b] * e) -> int
goal g : forall (x:refty[b,int]).
         forall (u:region[b]).
         (z =[int||] (f [unit||] (mk_2tuple [region[b],unit||] u tt))) ->
         ((ref_get [b,int||] u x) =[int||] 3)
